### ISSUE TITLE
Access violation (C0000005)

### ISSUE THREAD

**Issue by atifwaqar** [Starcounter/Starcounter#3608]
SC Version 2.2.1834
I just got my apps crashed. Here is what i have in my SC Logs (Source is Starcounter) :

```
Access violation (C0000005) detected at bmx.cpp:63 (Thread 17592).
    00007FFE21E88B9C KERNELBASE!
    00007FFE16DFA62C clr!
    00007FFE16DFA67A clr!
    00007FFE16DFA685 clr!
    00007FFDB7E55913 <unknown>!
```

This may not be related, but In logs i can see that 30 seconds before the "Access violation" exception I had the following exception (Source is Starcounter.Host):

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: Terrible situation: responseSizeBytes_ > estimatedNumBytes
   at Starcounter.Response.ConstructFromFields(Request req, Byte[] givenBuffer) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-283\Level1\src\Starcounter.Internal\Http\Response.cs:line 886
   at Starcounter.Request.SendResponse(Response resp, Byte[] serializationBuf) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-283\Level1\src\Starcounter.Internal\Http\Request.cs:line 1241
   at Starcounter.Internal.AppsBootstrapper.ProcessExternalRequest(Request req) in C:\TeamCity\TeamCity9\buildAgent\work\sc-develop-nightly-283\Level1\src\Starcounter.Apps.JsonPatch\AppsBootstrapper.cs:line 645
ParamName=Terrible situation: responseSizeBytes_ > estimatedNumBytes
HResult=-2146233086

```

**Comment by per-samuelsson** [comment]
@atifwaqar 

Thanks for reporting this.

```
Parameter name: Terrible situation: responseSizeBytes_ > estimatedNumBytes
```

I guess it's that kind of thing that could actually cause an AV.

@alemoi, any take on this?

**Comment by alemoi** [comment]
Thanks for reporting! Yes, its an uncaught CLR exception, that says that estimated response size is different from actual. I think it might be related to JSON serialization? @chrhol what do you think?

**Comment by chrhol** [comment]
> I think it might be related to JSON serialization? @chrhol what do you think?

Possibly, but the exception should be different if it was the serializer that caused it. There is the same check as the one logged in the end of serialization that verifies written vs estimated size, which also include some debuginfo about the json that caused the error.

**Comment by chrhol** [comment]
We could add some more information to this exception, like the URI for the incoming request and maybe some other stuff from the response to make it easier to guess where to start look.

**Comment by chrhol** [comment]
Lets assume as a starting point that this exception and the one described in #3604 are related. Depending on where it happens (since it's some kind of synchronization issue) you either get this or the other.

I will setup a test and try to provoke it and see what happens. And in parallel enhance the errrormesages to include more information to help pinpoint which resources that fails.

**Comment by chrhol** [comment]
The other issue had nothing to do with this one after all. I'm not sure what's caused this one. One thing that might happen (but maybe is a bit of a long shot) is if the response or serialized json is larger than 2 gb we get an overflow in the variable that holds the estimated size...and since there previously was no check for this, strange things might happen (like the exception above.)

I've added checks on the estimated size (that is calculated before anything is written) so that it doesn't exceed a fixed maximum size and tried to add additional information to the exceptions. Not sure how helpful they will be, but with check on estimated size, we should no longer have buffer overruns in these places that destroys memory, and causing the access violation.

These checks are added for json-serializer, generating jsonpatch and serializing responses.

I will push these changes to develop and pnext as soon as my personal build succeeds, so they will be included in tonights nightly build.

**Comment by chrhol** [comment]
Closing this in favor for #3735
